Feature/transaction history

drEfrati opened the pull request:
Payments History Feature

Commit:
Tx history feature

Commit:
Merge commit '1c5877b7d3f770bc86c86784d2770e52ff8be99d' into feature/transaction_history

# Conflicts:
#	kin-core/src/main/java/kin/core/AccountInfoRetriever.java
#	kin-core/src/main/java/kin/core/BlockchainEvents.java

Commit:
Remove unnecessary comments

Commit:
Remove TODOs and rename code from Transaction History to Payment History

mrsegev reviewed (approved):
Nice, just make the changes we talked about + update Readme file to reflect the new API.

yosriz commented on kin-core/src/main/java/kin/core/AccountInfoRetriever.java:
```diff
@@ -2,27 +2,40 @@
         this.kinAsset = kinAsset;
     }
 
+    void setBlockChainEvents(BlockchainEvents blockchainEvents) {
```
To be consistent with the current design - every dependency should be "injected" directly into the ctor (and then omit the null checks)

yosriz commented on kin-core/src/main/java/kin/core/BlockchainEvents.java:
```diff
@@ -135,27 +135,35 @@ public void onEvent(TransactionResponse transactionResponse) {
+        }
+    }
+
+    PaymentInfo getPaymentInfo(TransactionResponse transactionResponse) {
```
extract seems like a more appropriate name

yosriz commented on kin-core/src/main/java/kin/core/AccountInfoRetriever.java:
```diff
@@ -2,27 +2,40 @@
 
     private final Server server;
     private final KinAsset kinAsset;
+    private BlockchainEvents blockchainEvents;
```
can you extract the `TransactionResponse` -> `PaymentInfo` logic to a common class? it's not entirely clean to have ` AccountInfoRetriever` depends on unrelated `BlockchainEvents` class

yosriz commented on kin-core/src/main/java/kin/core/AccountInfoRetriever.java:
```diff
@@ -62,6 +75,74 @@ Balance getBalance(@NonNull String accountId) throws OperationFailedException {
+    private List<PaymentInfo> getPaymentsHistory(TransactionsRequestBuilder transactions, @NonNull String accountId) throws OperationFailedException {
+        List<PaymentInfo> payments = new ArrayList<>();
+        try {
+            Page<TransactionResponse> execute = transactions
```
execute -> response

yosriz commented on kin-core/src/main/java/kin/core/AccountInfoRetriever.java:
```diff
@@ -62,6 +75,74 @@ Balance getBalance(@NonNull String accountId) throws OperationFailedException {
+                    .forAccount(KeyPair.fromAccountId(accountId))
+                    .execute();
+            if (execute == null || execute.getRecords() == null) {
+                throw new AccountNotActivatedException(accountId);
```
why it's `AccountNotActivatedException` in this case?
what does null response mean? I would expect general exception (data cannot be retrieved) like in getBalance/sendTransaction

yosriz commented on kin-core/src/main/java/kin/core/AccountInfoRetriever.java:
```diff
@@ -62,6 +75,74 @@ Balance getBalance(@NonNull String accountId) throws OperationFailedException {
+                throw new OperationFailedException(httpError);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
rethrow this as `OperationFailedException`

yosriz commented on kin-core/src/main/java/kin/core/BlockchainEvents.java:
```diff
@@ -135,27 +135,35 @@ public void onEvent(TransactionResponse transactionResponse) {
             }
-
         }
+        // return a new payment info object if possible, otherwise return null.
```
seems a redundant comment

yosriz commented on kin-core/src/main/java/kin/core/KinAccount.java:
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
     Balance getBalanceSync() throws OperationFailedException;
 
+    /**
+     * Create {@link Request} for getting all the transactions payments history for the current account.
```
payment transactions

yosriz commented on kin-core/src/main/java/kin/core/KinAccount.java:
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
+     * Create {@link Request} for getting all the transactions payments history for the current account.
+     * <p> See {@link KinAccount#getPaymentsHistorySync()} for possibles errors</p>
+     *
+     * @return {@code Request<List<PaymentInfo>> } PaymentInfo - the payment information for a specific payment operation in a transaction
```
List of

yosriz replied:
the return value doc should specify it is the list of payments of this account.

yosriz commented on kin-core/src/main/java/kin/core/KinAccount.java:
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
+    /**
+     * For method description see {@link KinAccount#getPaymentsHistory()}
+     * @param requestParams an optional parameters to the request, such as accountId, cursor, limit, and order.
+     * Create {@link Request} for getting all the transactions payments history for a given account.
```
I don't think we should expose the possibility of giving an account id, as it doesn't match the current interfaces - `KinAccount` is for a specific address and it doesn't appropriate to give an option here to query other accounts.

yosriz commented on kin-core/src/main/java/kin/core/KinAccount.java:
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
+     *
+     * <p><b>Note:</b>     requestParams can have the next members:</p>
+     * <p><b>accountId</b> is optional, a string, if not given then using the current account if exist. It is represents the ID of an account.</p>
+     * <p><b>cursor:</b>   is optional, a paging token, specifying where to start returning records from. for example 12884905984.</p>
```
I'm not sure how token can be used here? what actually it is?
We should give the option to get this token from every payment o.w. it is not useful currently

yosriz commented on kin-core/src/main/java/kin/core/KinAccount.java:
```diff
@@ -104,6 +106,63 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
+    Request<List<PaymentInfo>> getPaymentsHistory(PaymentsHistoryRequestParams requestParams);
+
+    /**
+     * Get the list of the transactions payments history for the current account.
```
limit doc - maybe something like: default limit is 10, use `getPaymentsHistorySync(PaymentsHistoryRequestParams)` for providing a differnt limit

yosriz commented on kin-core/src/main/java/kin/core/KinAccountImpl.java:
```diff
@@ -57,6 +62,25 @@ public Balance getBalanceSync() throws OperationFailedException {
+    @NonNull
+    @Override
+    public List<PaymentInfo> getPaymentsHistorySync(PaymentsHistoryRequestParams requestParams) throws OperationFailedException {
+        String accountId = requestParams.getAccountId();
```
see comment above regarding specifying account id

yosriz commented on kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java:
```diff
@@ -0,0 +1,129 @@
+import android.os.Parcelable;
+
+/**
+ * Class which build optional params for the transaction payment history request.
```
Represents optional params for payment history request, see <add link to the method>

yosriz commented on kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java:
```diff
@@ -0,0 +1,129 @@
+        this.order = builder.order;
+    }
+
+    protected PaymentsHistoryRequestParams(Parcel in) {
```
why does it need to be parcelable?

yosriz commented on kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+
+    public static class PaymentsHistoryRequestParamsBuilder {
```
just Builder (`new PaymentsHistoryRequestParams.Builder`)

yosriz commented on kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java:
```diff
@@ -0,0 +1,129 @@
+        private int limit;                  //optional
+        private Order order;                //optional
+
+        public PaymentsHistoryRequestParamsBuilder account(String accountId) {
```
add "set" prefix to be consistent with other builders

yosriz replied:
add docs for setters

yosriz reviewed (changes requested):
:+1:
The most important comments are regarding the public APIs (the ability to change account id, the token)
Regarding tests, please see codecov report:
https://codecov.io/gh/kinecosystem/kin-core-android/commit/2534f9f47b3c99f2df00420ae8f1ddc117b28be4 
I'm not sure why codecov didn't put a comment on the PR like it does always (maybe you need to sign in?)

Commit:
1. Clear redundant code.
2. fix some little issues from code review
3. removing account and cursor(token) from the optional params and sample app UI
4. extract payment info methods to a common class and as a result AccountInfoRetriever doesn't depend anymore on BlockChainEvents

yosriz commented on kin-core/src/main/java/kin/core/KinAccount.java:
```diff
@@ -104,6 +106,57 @@ TransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDec
 
+    /**
+     * Create {@link Request} for getting list of payments history for the current account.
+     * The default is currently the last 10 payments, please use use {@link KinAccount#getPaymentsHistory((PaymentsHistoryRequestParams)} for providing a different limit.
```
"currently" is redundant, a suggestion: "The default limit is 10 payments".

yosriz replied:
redundant "use" (please use use)

yosriz commented on kin-core/src/main/java/kin/core/PaymentInfoHelper.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Helper class which handle the extraction of Payment Info from a transaction.
+ */
+public class PaymentInfoHelper {
```
should internal class with all the methods internal as well

yosriz commented on kin-core/src/main/java/kin/core/PaymentsHistoryRequestParams.java:
```diff
@@ -0,0 +1,83 @@
+        private int limit;                  //optional
+        private Order order;                //optional
+
+//        /**
```
dead code comment - should be deleted

yosriz reviewed (commented):
@drEfrati I've added small comments for small tweaks, other than that it's all good
